src/sundials_matrix_wrapper.cpp

```diff
@@ -349,37 +387,101 @@ void SUNMatrixWrapper::sparse_multiply(SUNMatrixWrapper *C,
-                                    "in A (" + std::to_string(nrows) + ") and "
-                                    "number of rows in C ("
+                                    "in A (" + std::to_string(nrows)
+                                    + ") and number of rows in C ("
                                     + std::to_string((int)C->rows()) + ")");
+    
+    if (C->columns() != B->columns())
+        throw std::invalid_argument("Dimension mismatch between number of cols "
+                                    "in B (" + std::to_string(B->columns())
+                                    + ") and  number of cols in C ("
+                                    + std::to_string((int)C->columns()) + ")");
 
     if (B->rows() != ncols)
-        throw std::invalid_argument("Dimension mismatch between number of rows "
+        throw std::invalid_argument("Dimension mismatch between number of cols "
                                     "in A (" + std::to_string(ncols)
-                                    + ") and number of cols in B ("
+                                    + ") and number of rows in B ("
                                     + std::to_string((int)B->rows()) + ")");
+    
+    if (ncols == 0 || nrows == 0 || B->columns() == 0)
+        return; // matrix will also have zero size
+    
+    if (nonzeros() == 0 || B->nonzeros() == 0)
+        return; //  nothing to multiply
+    
+
+    /* see https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/CSparse/Source/cs_multiply.c
+     * modified such that we don't need to use CSparse memory structure and can
+     * work with preallocated C. This should minimize number of necessary
+     * reallocations as we can assume that C doesn't change size.
+     */
+    
+    sunindextype nz = 0; // this keeps track of the nonzero index in C
+    auto Bx = B->data();
+    auto Bi = B->indexvals();
+    auto Bp = B->indexptrs();
+    
+    sunindextype j, p;
+    auto Cx = C->data();
+    auto Ci = C->indexvals();
+    auto Cp = C->indexptrs();
+    
+    sunindextype m = nrows;
+    sunindextype n = B->columns();
+    
+    auto w = std::vector<sunindextype>(m);
+    auto x = std::vector<realtype>(m);
+    
+    for (j = 0; j < n; j++)
+    {
+        Cp[j] = nz;                          /* column j of C starts here */
+        // CSparse reallocation happend when it _might_ have been necessary.
+        // We are a little bit more conservative here as we allow more
+        // reallocations on the first call but try to avoid them on later calls.
+
+        if ((Bp[j+1] > Bp[j]) && (nz + m > C->nonzero_space()))
+        {
+            C->reallocate(2*C->nonzero_space() + m);
+            Cx = C->data(); Ci = C->indexvals(); Cp = C->indexptrs();
+        }
+        for (p = Bp[j]; p < Bp[j+1]; p++) {
+            nz = scatter(Bi[p], Bx[p], w.data(), x.data(), j+1, C, nz);
+        }
+        for (p = Cp[j]; p < nz; p++)
+            Cx[p] = x.at(Ci[p]); // copy data to C
```

Comment:
check dimensions before, and don't use `at`? otherwise no vectorization possible.